### ISSUE TITLE
m4 package does not build with spack verion 0.13.0 on RHEL6 using spack-built gcc@7.2.0

### ISSUE THREAD

**Issue by bartlettroscoe** [spack/spack#13507]
CC: @fryeguy52 

Tag: snl-atdm-issue

I used spack to build gcc@7.2.0 using the native RHEL6 gcc@4.4.7 which gave the compilers.yaml entry:

```
- compiler:
    paths:
      cc: /home/rabartl/Spack.base4/atdm-spack/spack/opt/spack/linux-rhel6-core2/gcc-4.4.7/gcc-7.2.0-zzfkclh52gj6uzf4p64a3jbk2wo4id3i/bin/gcc
      cxx: /home/rabartl/Spack.base4/atdm-spack/spack/opt/spack/linux-rhel6-core2/gcc-4.4.7/gcc-7.2.0-zzfkclh52gj6uzf4p64a3jbk2wo4id3i/bin/g++
      f77: /home/rabartl/Spack.base4/atdm-spack/spack/opt/spack/linux-rhel6-core2/gcc-4.4.7/gcc-7.2.0-zzfkclh52gj6uzf4p64a3jbk2wo4id3i/bin/gfortran
      fc: /home/rabartl/Spack.base4/atdm-spack/spack/opt/spack/linux-rhel6-core2/gcc-4.4.7/gcc-7.2.0-zzfkclh52gj6uzf4p64a3jbk2wo4id3i/bin/gfortran
    operating_system: rhel6
    target: x86_64
    modules: []
    environment: {}
    extra_rpaths: []
    flags: {}
    spec: gcc@7.2.0
```

Then, using the file:

<details>

<summary><b>spack-yaml</b></summary>

```
# Spack env file for building ATDM spack tools packages
spack:
  specs: [
    snl-atdm-tools ,
    ]
  packages:
    all:
      compiler: [gcc@7.2.0]
  modules:
    tcl:
      hash_length: 0
      naming_scheme: spack-${PACKAGE}/${VERSION}
      whitelist: [python, perl, cmake, ninja-fortran, git, autoconf, automake, m4, libtool, tar, diffutils, util-macros]
      blacklist: ['%gcc@7.2.0']
      all:
        conflict:
          - ${PACKAGE}
        filter:
          environment_blacklist: [CPATH, LIBRARY_PATH]
        environment:
          set:
            ${PACKAGE}_ROOT: ${PREFIX}
      verbose: false

  include:
    - common_options.yaml
```
</details>

I tried `spack install` and it gave the error:

```
==> Installing m4
==> Searching for binary cache of m4
==> Finding buildcaches in /ascldap/users/rabartl/Spack.base4/atdm-spack/atdm-spack-tarball-mirror/build_cache
==> No binary for m4 found: installing from source
==> Using cached archive: /home/rabartl/Spack.base4/atdm-spack/spack/var/spack/cache/_source-cache/archive/ab/ab2633921a5cd38e48797bf5521ad259bdc4b979078034a3b790d7fec5493fab.tar.gz
==> Using cached archive: /home/rabartl/Spack.base4/atdm-spack/spack/var/spack/cache/_source-cache/archive/fc/fc9b61654a3ba1a8d6cd78ce087e7c96366c290bc8d2c299f09828d793b853c8
==> Staging archive: /tmp/rabartl/spack-stage/spack-stage-m4-1.4.18-npxsaugxhjz6girqfz7yzv4jfbneza6d/m4-1.4.18.tar.gz
==> Created stage in /tmp/rabartl/spack-stage/spack-stage-m4-1.4.18-npxsaugxhjz6girqfz7yzv4jfbneza6d
==> Applied patch /home/rabartl/Spack.base4/atdm-spack/spack/var/spack/repos/builtin/packages/m4/gnulib-pgi.patch
==> Applied patch https://src.fedoraproject.org/rpms/m4/raw/5d147168d4b93f38a4833f5dd1d650ad88af5a8a/f/m4-1.4.18-glibc-change-work-around.patch
==> Building m4 [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j20'

12 errors found in build log:
     563      CC       filenamecat.o
     564      CC       filenamecat-lgpl.o
     565      CC       fopen-safer.o
     566      CC       freading.o
     567      CC       getprogname.o
     568    /tmp/cch5UXCA.s: Assembler messages:
  >> 569    /tmp/cch5UXCA.s:60: Error: no such instruction: `shlx %rdi,%rax,%rd
            i'
  >> 570    /tmp/cch5UXCA.s:268: Error: no such instruction: `shlx %rdi,%rax,%r
            di'
  >> 571    /tmp/cch5UXCA.s:313: Error: no such instruction: `shlx %rdi,%rax,%r
            di'
  >> 572    make[3]: *** [c-ctype.o] Error 1
     573    make[3]: *** Waiting for unfinished jobs....
     574    /tmp/ccAJ3DAA.s: Assembler messages:
  >> 575    /tmp/ccAJ3DAA.s:767: Error: no such instruction: `shrx %eax,24(%rsi
            ),%eax'
  >> 576    /tmp/ccAJ3DAA.s:843: Error: no such instruction: `shrx %ecx,24(%rsi
            ),%ecx'
  >> 577    make[3]: *** [gl_avltree_oset.o] Error 1
     578    /tmp/cc1SkWzB.s: Assembler messages:
  >> 579    /tmp/cc1SkWzB.s:27: Error: no such instruction: `rorx $55,%rax,%rax
            '
  >> 580    make[3]: *** [clean-temp.o] Error 1
     581    make[3]: Leaving directory `/tmp/rabartl/spack-stage/spack-stage-m4
            -1.4.18-npxsaugxhjz6girqfz7yzv4jfbneza6d/spack-src/spack-build/lib'
  >> 582    make[2]: *** [all] Error 2
     583    make[2]: Leaving directory `/tmp/rabartl/spack-stage/spack-stage-m4
            -1.4.18-npxsaugxhjz6girqfz7yzv4jfbneza6d/spack-src/spack-build/lib'
  >> 584    make[1]: *** [all-recursive] Error 1
     585    make[1]: Leaving directory `/tmp/rabartl/spack-stage/spack-stage-m4
            -1.4.18-npxsaugxhjz6girqfz7yzv4jfbneza6d/spack-src/spack-build'
  >> 586    make: *** [all] Error 2

See build log for details:
  /tmp/rabartl/spack-stage/spack-stage-m4-1.4.18-npxsaugxhjz6girqfz7yzv4jfbneza6d/spack-build-out.txt


real	0m40.036s
user	0m26.897s
sys	0m10.030s
```

I am going to try to get around this by pointing to 'm4' and may other such packages already installed on the system.  (As I stated in #13198, the most fragile aspect of spack is building low-level packages that are already on most systems.)

**Comment by bartlettroscoe** [comment]
FYI: I could not get around this by pointing to 'm4' already on the system as I can't get that to work as described in #13516.

**Comment by bartlettroscoe** [comment]
@fryeguy52 says that to fix this, you need to have Spack build gcc@7.2.0 with BinUtils enabled and then this will work. 

@fryeguy52, can you please provide that spec to build gcc@7.2.0 with binutils to resolve?

**Comment by fryeguy52** [comment]
I hit this too.  @alalazo helped fix in #13362 the fix is to build gcc with this:
```
gcc@7.2.0 +binutils
```

**Comment by bartlettroscoe** [comment]
> `gcc@7.2.0 +binutils`

Okay, but that means more scripting.  Now we have to detect when the user wants to build a 'gcc' compiler and then we need to add the `+binutils` variant (because `llvm` does not have a `binutils` variant).

**Comment by alalazo** [comment]
@bartlettroscoe I don't think this is an issue strictly related to Spack. If you use the `binutils` found on the system then they must be able to translate to object code the assembly generated by the front-end. Usually this is an issue only on machines where, for any reason, an old operating system is in use. Let me know if you agree or not.

**Comment by bartlettroscoe** [comment]
>  I don't think this is an issue strictly related to Spack. If you use the `binutils` found on the system then they must be able to translate to object code the assembly generated by the front-end. Usually this is an issue only on machines where, for any reason, an old operating system is in use. Let me know if you agree or not.

@alalazo is there ever a reason not to build `binutils` with `gcc` if you are going to be building `gcc` with spack anyway?   If I want the Spack `gcc` to use a binutils already installed on the system, can't I just list that in a `packages.yaml` file?

The way I look at it is that we are trying to rely on Spack to create portable builds for software.  Can't the gcc/packagae.py file figure out that it needs to build a new binutils.   I mean, spack is not shy about building all kinds of low-level software by default (see #13198).  And many people's TPL build scripts end up building a new updated binutils so why would spack not build binutils by default when it builds GCC?

I know building software like this on many different systems is difficult.  I am just trying to figure out what Spack is supposed to be handing and what I need to handle on my own (and address in a growing scripting code base around spack).

**Comment by alalazo** [comment]
>  is there ever a reason not to build binutils with gcc if you are going to be building gcc with spack anyway? If I want the Spack gcc to use a binutils already installed on the system, can't I just list that in a packages.yaml file?

It's a fair point and I remember that a few years ago there were discussions on how to treat `binutils` in the `gcc` recipe.  Apparently the default was set to `False` in https://github.com/spack/spack/pull/4973 because of the lack of stability of the `gcc` recipe at the time. Maybe we should revisit at least the default value once we have public binary mirrors?

**Comment by bartlettroscoe** [comment]
> Maybe we should revisit at least the default value once we have public binary mirrors?

How does this relate to public binary mirrors?

**Comment by alalazo** [comment]
> How does this relate to public binary mirrors?

Just my personal opinion here, thinking defensively. If we switch the default after we have been able to generate binary packages for a number of reference systems we'll mitigate the risk of people not being able to build `gcc+binutils` like in #4973 and the push backs related to increased build times.

**Comment by alalazo** [comment]
Fixed by #14706
